src/Data/Foldable/Linear.hs

```diff
@@ -0,0 +1,160 @@
+    foldl1 f xs = fromMaybe (Prelude.errorWithoutStackTrace "foldl1: empty structure")
+                    (foldl mf Nothing xs)
+      where
+        mf :: Maybe a #-> a #-> Maybe a
+        mf Nothing y = Just y
+        mf (Just x) y = Just (f x y)
+
+    toList :: t a #-> [a]
+    toList = foldr (:) []
+
+    -- | This is likely not very useful and only here to preserve the symmetry
+    -- between 'Prelude.Foldable'.
+    null :: Consumable a => t a #-> Bool
+    null = foldr (\x a -> (x, a) `lseq` False) True
+
+    length :: Consumable a => t a #-> Int
+    length = foldl' (\c a -> a `lseq` c + 1) 0
+
+    -- | This is likely not very useful and only here to preserve the symmetry
+    -- between 'Prelude.Foldable'.
+    elem :: Prelude.Eq a => a -> t a #-> Bool
+    elem x = any (== x)
+
+    minimum :: (Dupable a, Prelude.Ord a) => t a #-> a
+    minimum = minimumBy compare
+
+    maximum :: (Dupable a, Prelude.Ord a) => t a #-> a
+    maximum = maximumBy compare
+
+instance {-# OVERLAPPABLE #-} (Foldable t, Consumable a) => Consumable (t a) where
+  consume = foldl' (flip lseq) ()
+
+instance Foldable [] where
+  foldr f z = \case
+    [] -> z
+    x:xs -> f x (foldr f z xs)
+
+instance Foldable Maybe where
+  foldMap _ Nothing = mempty
+  foldMap f (Just s) = f s
+
+instance Foldable NonEmpty where
+  foldMap f (x:|xs) = f x <> foldMap f xs
+
+instance Consumable a => Foldable (Either a) where
+  foldMap f (Left l) = l `lseq` mempty
+  foldMap f (Right r) = f r
+
+maximumBy :: (Foldable t, Dupable a) => (a #-> a #-> Ordering) -> t a #-> a
+maximumBy cmp = foldl1 (\x y ->
+  dup (x, y) & \case
+    ((x1, y1), (x2, y2)) -> cmp x1 y1 & \case
+      GT -> y2 `lseq` x2
+      EQ -> x2 `lseq` y2
+      LT -> x2 `lseq` y2
+  )
+
+minimumBy :: (Foldable t, Dupable a) => (a #-> a #-> Ordering) -> t a #-> a
+minimumBy cmp = maximumBy (\x y -> cmp x y & \case
+    GT -> LT
+    EQ -> EQ
+    LT -> GT
+  )
+
+-- | __NOTE:__ This does not short-circuit, and always consumes the
+-- entire container.
+any :: Foldable t => (a #-> Bool) -> t a #-> Bool
+any p = foldl' (\b a -> b || p a) False
+
+-- | __NOTE:__ This does not short-circuit, and always consumes the
+-- entire container.
+all :: Foldable t => (a #-> Bool) -> t a #-> Bool
+all p = foldl' (\b a -> b && p a) True
+
+-- | __NOTE:__ This does not short-circuit, and always consumes the
+-- entire container.
+and :: Foldable t => t Bool #-> Bool
+and = foldl' (&&) True
+
+-- | __NOTE:__ This does not short-circuit, and always consumes the
+-- entire container.
+or :: Foldable t => t Bool #-> Bool
+or = foldl' (||) False
+
+traverse_ :: (Foldable t, Applicative f, Consumable b)
+          => (a #-> f b) -> t a #-> f ()
+traverse_ f = foldr (\x k -> f x *> k) (pure ())
+
+sequence_ :: (Foldable t, Applicative f, Consumable a)
+          => t (f a) #-> f ()
+sequence_ = traverse_ id
+
+foldMapDefault :: (Traversable t, Monoid m, Consumable (t ())) => (a #-> m) -> t a #-> m
```

Comment:
I think documenting a bit of the discussion of `Consumable t ()` is good here. Like how not all `Traversible` things are `Foldable` and this is in part seen by the fact that not all `Traversible` things `t` have a `Consumable (t ())` instance.